lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystem.java

```diff
@@ -202,4 +211,34 @@ private static void validateS3Location(Location location)
     {
         new S3Location(location);
     }
+
+    private boolean shouldReadObject(S3Object object)
+    {
+        return switch (context.s3ObjectStorageClassFilter()) {
+            case READ_ALL -> true;
+            case SKIP_ALL_GLACIER -> !isGlacierObject(object);
+            case READ_RESTORED_GLACIER_OBJECTS -> !isGlacierObject(object) || isCompletedRestoredObject(object);
+        };
+    }
+
+    private static boolean isCompletedRestoredObject(S3Object object)
+    {
+        /* There are 3 cases for the restore status:
+         *
+         * 1. The object is not restored, and has not been requested to be restored. We should have a null
+         * value for restoreStatus
+         * 2. The object is in the process of being restored. isRestoreInProgress will be true, but restoreExpiryDate
+         * will be null.
+         * 3. The object has completed restore. isRestoreInProgress will be false, restoreExpiryDate will be some date.
+         *
+         * Since we only care about distinguishing when it is case 3, we can just check restoreExpiryDate for a non-null value. */
+        return Optional.ofNullable(object.restoreStatus())
+                .map(RestoreStatus::restoreExpiryDate)
+                .isPresent();
+    }
+
+    private static boolean isGlacierObject(S3Object object)
+    {
+        return GLACIER_STORAGE_CLASSES.contains(object.storageClass());
```

Comment:
Let's simplify this to
```java
return (object.storageClass() == GLACIER) || (object.storageClass() == DEEP_ARCHIVE);
```
Or we could do
```java
return switch (object.storageClass()) {
    case GLACIER, DEEP_ARCHIVE -> true;
    default -> false;
};

Reply:
Will do